File: data/skeletons/JavaFastPFOR/src/main/me/lemire/integercompression/DeltaZigzagEncoding.py
from __future__ import annotations

# Imports Begin
import typing
from typing import *
import io

# Imports End


class Context:

    # Class Fields Begin
    contextValue: int = None
    # Class Fields End

    # Class Methods Begin
    def getContextValue(self) -> int:
        pass

    def setContextValue(self, contextValue: int) -> None:
        pass

    def __init__(self, contextValue: int) -> None:
        pass

    # Class Methods End


class Decoder(Context):

    # Class Fields Begin
    # Class Fields End

    # Class Methods Begin
    def decodeArray2(self, src: typing.List[int]) -> typing.List[int]:
        pass

    def decodeArray1(
        self, src: typing.List[int], offset: int, length: int
    ) -> typing.List[int]:
        pass

    def decodeArray0(
        self,
        src: typing.List[int],
        srcoff: int,
        length: int,
        dst: typing.List[int],
        dstoff: int,
    ) -> typing.List[int]:
        pass

    def decodeInt(self, value: int) -> int:
        pass

    def __init__(self, contextValue: int) -> None:
        pass

    # Class Methods End


class Encoder(Context):

    # Class Fields Begin
    # Class Fields End

    # Class Methods Begin
    def encodeArray3(self, src: typing.List[int]) -> typing.List[int]:
        pass

    def encodeArray2(
        self, src: typing.List[int], offset: int, length: int
    ) -> typing.List[int]:
        pass

    def encodeArray1(
        self, src: typing.List[int], srcoff: int, length: int, dst: typing.List[int]
    ) -> typing.List[int]:
        pass

    def encodeArray0(
        self,
        src: typing.List[int],
        srcoff: int,
        length: int,
        dst: typing.List[int],
        dstoff: int,
    ) -> typing.List[int]:
        pass

    def encodeInt(self, value: int) -> int:
        pass

    def __init__(self, contextValue: int) -> None:
        pass

    # Class Methods End


class DeltaZigzagEncoding:

    # Class Fields Begin
    # Class Fields End

    # Class Methods Begin
    # Class Methods End

    pass

File: java_projects/cleaned_final_projects_decomposed_tests/JavaFastPFOR/src/main/java/me/lemire/integercompression/DeltaZigzagEncoding.java
/*
 * This code is released under the
 * Apache License Version 2.0 http://www.apache.org/licenses/.
 */
package me.lemire.integercompression;

/**
 * Delta+Zigzag Encoding.
 * 
 * @author MURAOKA Taro http://github.com/koron
 */
public final class DeltaZigzagEncoding {

        static class Context {
                int contextValue;

                Context(int contextValue) {
                        this.contextValue = contextValue;
                }

                void setContextValue(int contextValue) {
                        this.contextValue = contextValue;
                }

                int getContextValue() {
                        return this.contextValue;
                }
        }

        static class Encoder extends Context {
                Encoder(int contextValue) {
                        super(contextValue);
                }

                int encodeInt(int value) {
                        int n = value - this.contextValue;
                        this.contextValue = value;
                        return (n << 1) ^ (n >> 31);
                }

                int[] encodeArray0(int[] src, int srcoff, int length,
                                   int[] dst, int dstoff) {
                        for (int i = 0; i < length; ++i) {
                                dst[dstoff + i] = encodeInt(src[srcoff + i]);
                        }
                        return dst;
                }

                int[] encodeArray1(int[] src, int srcoff, int length,
                                   int[] dst) {
                        return encodeArray0(src, srcoff, length, dst, 0);
                }

                int[] encodeArray2(int[] src, int offset, int length) {
                        return encodeArray0(src, offset, length,
                                new int[length], 0);
                }

                int[] encodeArray3(int[] src) {
                        return encodeArray0(src, 0, src.length,
                                new int[src.length], 0);
                }
        }

        static class Decoder extends Context {
                Decoder(int contextValue) {
                        super(contextValue);
                }

                int decodeInt(int value) {
                        int n = (value >>> 1) ^ ((value << 31) >> 31);
                        n += this.contextValue;
                        this.contextValue = n;
                        return n;
                }

                int[] decodeArray0(int[] src, int srcoff, int length,
                                   int[] dst, int dstoff) {
                        for (int i = 0; i < length; ++i) {
                                dst[dstoff + i] = decodeInt(src[srcoff + i]);
                        }
                        return dst;
                }

                int[] decodeArray1(int[] src, int offset, int length) {
                        return decodeArray0(src, offset, length,
                                new int[length], 0);
                }

                int[] decodeArray2(int[] src) {
                        return decodeArray1(src, 0, src.length);
                }
        }
}
